### PR TITLE
refactor: move recharts scale code into recharts repository

### PR DESCRIPTION
**Pull request by ForestLinSen**
## Description

Edit: This PR will now focus only on the code movement from `recharts-scale` as per the discussion

~~This PR is for an issue where minor changes in data could affect the Y-axis domain significantly(an example mentioned in the issue #4020 [here](https://codesandbox.io/p/sandbox/bar-chart-stacked-by-sign-forked-3hnrht?file=%2Fsrc%2FApp.tsx%3A22%2C17)). Various parameters impact this, but a relatively straightforward solution, after quite a few experiments, is to adjust the `stepRatioScale`.~~

~~This value was previously hardcoded to 0.05, it now supports adjustments among 0.01, 0.03, or 0.05. A value closer to 0 makes the Y-axis domain closely match the data's maximum value (it cannot be 0 to avoid division by zero errors). Higher values can cause the Y-axis to scale to excessively high values.~~

~~Since this change needs to modify the code in `recharts-scale`, after discussion with Coltin, I've moved the necessary code parts for this PR from `recharts-scale` directly into this project to make it easier to test and manage.~~

## Related Issue

[Weired behaviour when calculating yAxis domain: ["auto", "auto"]](https://github.com/recharts/recharts/issues/4020), 

## How Has This Been Tested?

Locally tested and unit test

## Screenshots (if appropriate):

https://github.com/recharts/recharts/assets/41566276/a1329272-e60b-4350-babf-308911c8bfc7

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have added a storybook story or extended an existing story to show my changes

### CODE REVIEW

File: src/chart/generateCategoricalChart.tsx
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1232]: does this comment need removed?

File: src/chart/generateCategoricalChart.tsx
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:13]: getNickTickValues? Do you mean getNiceTickValues?

File: storybook/stories/API/chart/BarChart.stories.tsx
Comment (ckifer) [inline storybook/stories/API/chart/BarChart.stories.tsx:50]: nit: can remove the log here

File: storybook/stories/API/props/ChartProps.ts
Comment (ckifer) [inline storybook/stories/API/props/ChartProps.ts:185]: I'm not sure if this is enough description for people to understand what this is and what it accomplishes

File: test/util/ChartUtils.spec.tsx
Comment (ckifer) [inline test/util/ChartUtils.spec.tsx:353]: nit: I can't understand from the title what this test is doing

**Review by ckifer — CHANGES_REQUESTED** [review]
Appreciate you digging into this. Looks like a few things need fixed/updated and the build is breaking. I'd really like to see if this affects the storybook diffs at all before merging, I can kick that off at some point.

Nice find, but does giving the user the ability to tune this parameter really solve the problem? By default the same thing will happen as before. Now the user has to understand this stepRatioControl prop and really know what it is doing... I barely know what it is doing 😅. There doesn't seem to be any better default we can change this to rather than making a new prop?

**Comment by ForestLinSen** [comment]
> Appreciate you digging into this. Looks like a few things need fixed/updated and the build is breaking. I'd really like to see if this affects the storybook diffs at all before merging, I can kick that off at some point.
> 
> Nice find, but does giving the user the ability to tune this parameter really solve the problem? By default the same thing will happen as before. Now the user has to understand this stepRatioControl prop and really know what it is doing... I barely know what it is doing 😅. There doesn't seem to be any better default we can change this to rather than making a new prop?

thanks for the comments, still need to fix a few things later, but just want to discuss the default value first - I'm a little bit hesitant to change the default value, the number was set to 0.05 because we wanted to have nice numbers, like [100, 200, 300] instead of [80, 180, 280] (the values may be different but you get the point). It's only in cases where the data is large that minor changes cause significant shifts in the y domain, for smaller values it generally works well. So I'm inclined to set this as the default number, and give users the option to adjust it should they encounter any edge cases. What are your thoughts on this?

**Comment by ckifer** [comment]
Having the option to change the behavior is better than not having the option to change it so to me, all good, it's a good increment. Ideally we would find a more intuitive way to adjust it - this number seems relatively arbitrary and isn't something easy to comprehend without knowing implementation details. 

No matter what we need the documentation to be crystal clear as to what this actually changes

**Review by ckifer — COMMENTED** [review]
I think this is a great start (compared to JS code and no tests in the other repo). @PavelVanecek any thoughts or concerns here?

File: test/util/scale/utils.spec.ts
Comment (nikolasrieble) [inline test/util/scale/utils.spec.ts:5]: 💭 Love the tests, I am a bit sceptical as to whether we need these helper functions, and whether we could just get by with a simpler solution that just might be more verbose - but easier to read and maintain.

**Review by nikolasrieble — APPROVED** [review]
Pretty great work. It adds value for users, and is nicely documented.

I would like a separate example story just for this parameter. I.e. a chart with multiple yAxis, the same dataset, just showing how the values differ. Not a requirement. 

Code wise, I wonder whether some of the code could be simplified, but also not blocking.

File: src/util/scale/getNiceTickValues.ts
Comment (ckifer) [reply src/util/scale/getNiceTickValues.ts:127]: this isn't his code, this was copied from `recharts-scale` and the original author from that repo

File: test/util/scale/utils.spec.ts
Comment (ckifer) [reply test/util/scale/utils.spec.ts:5]: These were all brought over from recharts-scale

Your concern might be true (we don't need these), but we should get more test coverage first to ensure nothing breaks

File: src/util/scale/util/arithmetic.ts
Comment (ckifer) [reply src/util/scale/util/arithmetic.ts:45]: yeah this is weird, but also a recharts-scale artifact https://github.com/recharts/recharts-scale/blob/master/src/util/arithmetic.js#L46

File: src/util/scale/util/utils.ts
Comment (ckifer) [reply src/util/scale/util/utils.ts:3]: recharts scale reference https://github.com/recharts/recharts-scale/blob/master/src/util/utils.js#L4

**Comment by nikolasrieble** [comment]
Just sharing because I forgot to mention: The best way to do this change would have been a two step approach: 
(1) First PR only to move the recharts-scale into the repo. Add tests, etc.
(2) Second PR to build a new feature. 

With that approach, review would be quicker, reviewers would feel more comfortable.
Again, this is great, thank you. Just want to share feedback on how it could be even better ❤️

**Comment by ForestLinSen** [comment]
> Just sharing because I forgot to mention: The best way to do this change would have been a two step approach: (1) First PR only to move the recharts-scale into the repo. Add tests, etc. (2) Second PR to build a new feature.
> 
> With that approach, review would be quicker, reviewers would feel more comfortable. Again, this is great, thank you. Just want to share feedback on how it could be even better ❤️

Thanks, that's a great suggestion. For now, I haven't moved all the code from rechart-scale, just the parts necessary for ticket computation. It might be worth creating another PR for all the remaining code and adding tests for it

**Comment by ckifer** [comment]
Any concerns before merge @PavelVanecek?

File: package.json
Comment (PavelVanecek) [inline package.json:80]: If we're merging recharts-scale inside the main package, can we remove the dependency now?

File: test/util/scale/getNiceTickValues.spec.ts
Comment (PavelVanecek) [inline test/util/scale/getNiceTickValues.spec.ts:1]: How does this compare to the previous tests? https://github.com/recharts/recharts-scale/blob/master/test/ticks/getNiceTickValues.test.js based on a quick scroll it looks like it used to have more? Did you decide to omit some of them?

File: storybook/stories/API/props/ChartProps.ts
Comment (PavelVanecek) [inline storybook/stories/API/props/ChartProps.ts:189]: What does this parameter do that `domain` doesn't?

Comment (ForestLinSen) [reply storybook/stories/API/props/ChartProps.ts:189]: Hi as the original [issue](https://github.com/recharts/recharts/issues/4020) mentioned it's for addressing the problem when the domain is set to ["auto", "auto"]

File: test/util/scale/getNiceTickValues.spec.ts
Comment (ForestLinSen) [reply test/util/scale/getNiceTickValues.spec.ts:1]: Good point, will add more tests based on this

File: storybook/stories/API/props/ChartProps.ts
Comment (PavelVanecek) [reply storybook/stories/API/props/ChartProps.ts:189]: Yeah, auto does what auto does, people who are not happy with the default auto behavior can go and provide their own behavior. Why add another prop that does the same thing?

File: package.json
Comment (ForestLinSen) [reply package.json:80]: didn't add all the code from `recharts-scale` just the parts necessary for the changes for now

File: storybook/stories/API/props/ChartProps.ts
Comment (ForestLinSen) [reply storybook/stories/API/props/ChartProps.ts:189]: I think one difference when setting to auto is for the "nice" value. Like, for a max value of 3374, 'dataMax + 10%' would be 3711 and auto with step ratio control can yield values like 3600 or 4000

Comment (PavelVanecek) [reply storybook/stories/API/props/ChartProps.ts:189]: The trouble is, what's nice or not is subjective.

In my opinion the real solution is to provide in the 'domain' prop the values that work for you in your chart and feel nice for you.

Comment (ckifer) [reply storybook/stories/API/props/ChartProps.ts:189]: This isn't about the domain, it's about what tick values the domain produce. Domain is still needed and these two properties do not do the same thing.

Comment (PavelVanecek) [reply storybook/stories/API/props/ChartProps.ts:189]: Aha! Can we find a different property that achieves the same and is more relatable? For example tick count, or ticks padding?

Comment (ckifer) [reply storybook/stories/API/props/ChartProps.ts:189]: Ok maybe I misspoke slightly, it controls what the values of the domain will be (the min, the max, and the in between round)m numbers) when the domain is set to auto. 

I agree that this property isn't relatable. I don't agree that the subjectivity of "nice" should prevent us from taking a subjective opinion on nice. We should have good defaults. We don't need perfect defaults but they should be solid - the behavior described in the linked issue is odd and worth changing. There might be a better way to change it than this, but I still think it's worth changing

Comment (PavelVanecek) [reply storybook/stories/API/props/ChartProps.ts:189]: I would prefer to just change the default then instead. We can release that as a breaking change in 3.x.

I'm expecting we won't be able to explain the new property and people will struggle to use it. And because it's a part of the public API then it will be impossible to change.

Comment (PavelVanecek) [reply storybook/stories/API/props/ChartProps.ts:189]: > controls what the values of the domain will be:
I agree we need some sort of control and customization for this

> the min, the max
This is the `domain` prop

> and the in between round)m numbers
Looking at recharts.org I can see that there already is `tickCount` prop on YAxis. There's also `ticks` and `minTickGap` which are easy to picture and understand the relationship between input and output. (`ticks={3}` means there will be 3 ticks.)

Then there's `scale` which is somewhat more abstract but on the other hand standardized in other charting libraries and experiences.

I don't think we need a new property and I am not at all convinced that values `0.01 | 0.03 | 0.05` are good values that communicate what they do, externally.

Comment (ckifer) [reply storybook/stories/API/props/ChartProps.ts:189]: I agree that this is hard to explain/understand. 

The properties you mention don't do the same thing as this though. 

Sure the min and max are set with domain, but when you set your own domain the generated ticks become non-round numbers like 554, 239, 38472847274, etc. 

When auto is used the ticks become 100, 250, 500, 750, 1000, etc. and they're still evenly spaced (if you specified your own domain, ticks, and tick count they would not be). 

The only way you could get the same result yourself is if you did some calculation to determine you domain using your data beforehand and make it always easily divisible by the number of ticks you have. Doing that is a lot tougher than just having recharts do it by default (and I'm not convinced that doing what I just said works 😅)

Just saying this to emphasize that it's something we should at least take some responsibility for, at some point. Let's wait until after 3.x and revisit unless we can come up with a less complicated way of improving this

Comment (ckifer) [reply storybook/stories/API/props/ChartProps.ts:189]: We can still certainly keep the recharts-scale movement

**Comment by ckifer** [comment]
@ForestLinSen would you be able to remove the code that is specific to `stepRatioControl` and only treat this PR as recharts-scale movement? That was way we can unblock what most of this work was and merge it and keep the discussion going about stepRatioControl or other solutions to the tick problem. 

Feel free to do it here or close this and make a new PR

**Comment by ForestLinSen** [comment]
@ckifer Sure, I have removed the code related to `stepRatioControl` and will make a new PR about the parameter after this one

**Pull request merged**
